scrapy/pipelines/files.py

```diff
@@ -480,9 +480,38 @@ def _warn():
         ## end of deprecation warning block
 
         media_guid = hashlib.sha1(to_bytes(url)).hexdigest()  # change to request.url after deprecation
-        media_ext = os.path.splitext(url)[1]  # change to request.url after deprecation
+        media_ext = self._get_file_extension(url, response or Response("")) # change to request.url after deprecation
         return 'full/%s%s' % (media_guid, media_ext)
 
+    def _get_file_extension(self, url , response):
+        """Helper function to obtain the filename from the response headers, falling back to the original request path
+
+            This function will look at the standard 'Content-Type' and 'Content-Disposition' HTTP response headers
+            and will try to obtain the file extension from it. If they are not present, or does not contain valid
+            that information, this function will return the file extension on original request path.
+
+            IE:
+                * http://www.example.com/image.jpg yields '.jpg'
+                * http://www.example.com/media.do?filename=image.jpg yields '.do'
+                * http://www.example.com/image yields an empty string
+        """
+
+        filename = urlparse(url).path[1:]
+        content_type = response.headers.get('Content-Type', b'')
+        content_disposition = response.headers.get('Content-Disposition', b'')
+
+        if content_type.startswith(b'application/octet-stream '):
+            return content_type[25:]
```

Comment:
What will happen if `Content-Type` is `'application/octet-stream /../../../../../asdf'`?